tokio/src/sync/semaphore.rs

```diff
@@ -1,1070 +1,183 @@
-            };
-        }
-
-        loop {
-            let mut next = curr;
-
-            if curr.is_closed() {
-                undo_strong!();
-                return Ready(Err(AcquireError::closed()));
-            }
-
-            if !next.acquire_permit(&self.stub) {
-                debug_assert!(curr.waiter().is_some());
-
-                if maybe_strong.is_none() {
-                    if let Some((ref mut cx, ref mut permit)) = permit {
-                        // Get the Sender's waiter node, or initialize one
-                        let waiter = permit
-                            .waiter
-                            .get_or_insert_with(|| Arc::new(WaiterNode::new()));
-
-                        waiter.register(cx);
-
-                        if !waiter.to_queued_waiting() {
-                            // The node is alrady queued, there is no further work
-                            // to do.
-                            return Pending;
-                        }
-
-                        maybe_strong = Some(WaiterNode::into_non_null(waiter.clone()));
-                    } else {
-                        // If no `waiter`, then the task is not registered and there
-                        // is no further work to do.
-                        return Pending;
-                    }
-                }
-
-                next.set_waiter(maybe_strong.unwrap());
-            }
-
-            debug_assert_ne!(curr.0, 0);
-            debug_assert_ne!(next.0, 0);
-
-            match next.compare_exchange(&self.state, curr, AcqRel, Acquire) {
-                Ok(_) => {
-                    match curr.waiter() {
-                        Some(prev_waiter) => {
-                            let waiter = maybe_strong.unwrap();
-
-                            // Finish pushing
-                            unsafe {
-                                prev_waiter.as_ref().next.store(waiter.as_ptr(), Release);
-                            }
-
-                            return Pending;
-                        }
-                        None => {
-                            undo_strong!();
-
-                            return Ready(Ok(()));
-                        }
-                    }
-                }
-                Err(actual) => {
-                    curr = actual;
-                }
-            }
-        }
+    pub fn available_permits(&self) -> usize {
+        self.ll_sem.available_permits()
     }
 
     /// Close the semaphore. This prevents the semaphore from issuing new
     /// permits and notifies all pending waiters.
-    pub(crate) fn close(&self) {
-        // Acquire the `rx_lock`, setting the "closed" flag on the lock.
-        let prev = self.rx_lock.fetch_or(1, AcqRel);
-
-        if prev != 0 {
-            // Another thread has the lock and will be responsible for notifying
-            // pending waiters.
-            return;
-        }
-
-        self.add_permits_locked(0, true);
+    pub fn close(&self) {
```

Comment:
I find closing a `Semaphore` a bit akward. While it can certainly be enabled using this underlying primitive, the purpose is a bit unclear. And any other framework doesn't allow it.

Reply:
I have never seen this kind of API before either. The POSIX semaphore API looks a lot different: `sem_wait`, `sem_trywait`, `sem_timedwait`, `sem_getvalue`, `sem_post`, ...

Maybe calling that thing `Semaphore` is wrong altogether. How about a different name like `Multilock` or `Multipass'?

I quite like the idea of being able to `close` it. It adds symmetry with `channel` where closing the `Receiver` also makes the `Sender::try_send` fail with `TrySendErrro::Closed`.

Reply:
I think we can remove `close` from this type here. It was mostly added for `mpsc`. If we hit a use case, we can revisit it for 0.3.

Reply:
I found that one to be quite useful when implementing some kind of "shutdown", but I'm fine with removing it from the API.

Reply:
Just wondering: Shall I remove both `remove_permits` and `try_remove_permits` from the PR?

Reply:
When removing the `close` function the case where `acquire` returns an `Err` should no longer happen so the `AcquireError` will go away. The `TryAcquireError` can also be replaced by an `Option` as there is now only one possible reason why no permit was returned.